### ISSUE TITLE
Rule proposal: forbid unnecessary assignment of parameter properties

### ISSUE THREAD

**Issue by EvgenyOrekhov** [typescript-eslint/typescript-eslint#7045]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

My rule would check that there are no unnecessary assignments of parameter properties in constructors.

### Fail Cases

```typescript
class Foo {
  constructor(public bar: string) {
    this.bar = bar; // <-- warn here
  }
}

class Foo2 {
  baz: string;
  constructor(public bar: string, baz: string) {
    this.bar = bar; // <-- warn here
    this.baz = baz;
  }
}
```

### Pass Cases

```typescript
class Foo {
  constructor(public bar: string) {}
}

class Foo2 {
  baz: string;
  constructor(public bar: string, baz: string) {
    this.baz = baz;
  }
}
```

### Additional Info

You can see in [this TypeScript playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ1rHgOwgFwCcBXYPeAgCgAciAjEAS2GlrAIC5p8CHsBzAJQp0GaHgAWDCADpWBaAF4WbANwiAvqk1A) that assignment of parameter properties produces duplicate assignment:

<img width="688" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/8045060/a5161970-d0cb-4868-8126-0535d0b4fcd5">

**Comment by JoshuaKGoldberg** [comment]
Aha I've definitely made this mistake myself. +1!

**Comment by undsoft** [comment]
I can take a stab at it.

**Comment by yeonjuan** [comment]
Hi @JoshuaKGoldberg I've started implementing this rule and have a question about https://github.com/typescript-eslint/typescript-eslint/pull/7089#discussion_r1257382039 in the previous PR.

Does this comment mean that the expectation is to only check the assignment directly below the function body? 
If so, it would allow the following cases.

```ts
class Foo {
  constructor(public foo: string, bar: boolean) {
      if (bar) {
          this.foo = foo;
       }
  }
}
```

**Comment by bradzacher** [comment]
@yeonjuan that specific case would be invalid as you're assigning the initial value to itself and it's not within a function scope.

The edge-case that comes to mind would be this code:
```ts
class Foo {
  resetFoo: () => void;

  constructor(public foo: string) {
    this.resetFoo = () => {
      this.foo = foo;
    };
  }
}

const foo = new Foo("a");
console.log(foo.foo); // -> "a"
foo.foo = "b";
console.log(foo.foo); // -> "b"
foo.resetFoo();
console.log(foo.foo); // -> "a"
```

But there's also this edge case:
```ts
class Foo {
  constructor(public foo: string, bar: string) {
    if (bar) {
      this.foo = bar;
    } else {
      this.foo = foo;
    }
  }
}
```
Should we error here and enforce that the code drops the `else` branch?

This further refinement of that case probably needs some thought too:
```ts
class Foo {
  constructor(public foo: number, bar?: number) {
    if (bar) {
      this.foo = foo - bar;
    }

    if (this.foo < 0) {
      this.foo = foo;
    }
  }
}
```
Should this code error? We could handle this by tracking assignments?

It's worth mentioning that it's also ***completely valid*** for us to declare all of this code as a known edge case that we don't want to handle for now. I.e. declare a failing test with a comment saying as much.
The likely scenario is that these usecases are purely academic and so no user would ever hit them or complain.

**Comment by kirkwaiblinger** [comment]
IMO, the more errors the better. If you're reassigning a parameter property... you should just make it a normal property and assign it with a normal parameter anyway, especially if you have something so weird that you're deliberately reassigning it to itself. So, I personally would break on the side of flagging rather than not flagging in the case of any ambiguities, such as the ones above.

Another proposal, though, that extends the spirit of this rule, could be "no using the parameter at all; only permit access via `this.foo`". 

```ts
class Foo {
  resetFoo: () => void;

  constructor(public foo: string) {
    console.log(foo)
                ^^^ 
    /* error, use `this.foo` to access `foo`, or explicitly create a variable with its initial value */
  }
}
```

That forces you to be explicit about capturing the initial value of property into an appropriately named variable if that's what you actually want. IMO, the below code is unambiguously better than using the `foo` parameter to capture an initial value anyway. 

```ts
class Foo {
  resetFoo: () => void;

  constructor(public foo: string) {
    const initialFoo = this.foo;
    this.resetFoo = () => {
      this.foo = initialFoo;
    };
  }
}
```
I'd wager most uses of `foo` without `this.foo` in this context are written without taking into consideration that the two values could differ, either because
1. easy accidental oversight
2. Muscle memory coming from c++/java/C# where `this.foo` is equivalent to `foo`, and including `this` is often considered bad style (an opinion I disagree with). In most cases, you can rely on the TS compiler to rescue you, by giving an error and saying "did you mean `this.foo`", but not here, due to there being an identifier of the same name but potentially different value in scope, from a single declaration 🤮 

Along the lines of point 2 above, this proposal has the added benefit of simultaneously preventing errors resulting from writing `foo = bar` when `this.foo = bar` was meant, a situation, that again, the compiler won't help you with, and should probably get flagged in its own lint rule if it doesn't make it into this one.

**Comment by kirkwaiblinger** [comment]
Unfortunately, there's one footgun I can think of with the above approach, which is that depending on your tsconfig _target_ setting and _usedDefineForClassFields_ setting, `this.foo` may not equal `foo`, even in the first line of the constructor.

```ts
class RunsCodeBeforeConstructor {
  constructor(public foo: number = 0) {
    console.log(`parameter and property are equal: ${this.foo === foo}`);
  }

  // @ts-ignore: Whether this is considered an error depends on useDefineForClassFields and target
  mayRunBeforeConstructorSemiChecked = this.foo += 10;

  // never an error
  mayRunBeforeConstructor1NeverChecked = (() => this.foo += 10)();
}
```

([playground link](https://www.typescriptlang.org/play?target=99&useDefineForClassFields=true#code/MYGwhgzhAEBKCuA7CBhA9gEwKYCEsDM0AnLdZAFyPmHOOgG8AoaaYNCqm4gCgAd4ARiACWwaITQAuaIngBbAViLQAvNAAMASgbMWrdhDQgsAOhBoA5twAGvMETBys5JdDCIM0XkTS8l5AE83EmgsAEd4MBBpABJ6cgALYQgTCVUVNQkAX2tNAG5dLMZdAHoS6AABcggAWmELRGIsaQB1BOd25UTk6B62ZGFsEk93UKIfZWw-Dxh2aHgILAARAmFELAAxYhRwKA3hLBAMGHdPcnsLZ105MACERDxCEjIISmpaIgBlLDlhFHbgABrLCeNTdFJpADUagAjOoCqVyusAG6uUZKCbXW73R5NF5vLhEGEAOSwqKI-ywQJBqmg3G42hUAD5oODUmg0NBodA4ZoGQUisV1gB3OBIVCYXAEPEGAkfflAA))

So we might need to use caution if implying to users that `const initialFoo = this.foo` is always guaranteed to work correctly. Or this could be a documentation candidate for a `:::warning` block, or a "When not to use this rule" section.

_This impacts users who have target >= ECMA2022 and have explicitly set `useDefineForClassFields: false`, or users who have target < ECMA 2022 (unless they have opted in to `useDefineForClassFields: true`)_

FWIW - I only know about this because I ran into a bug trying to upgrade a project to target: 2022. So it's a real thing. _BUT_, even that bug (which had to do with required property initializer side effects) was far less esoteric IMO than mutating a constructor property.
*** 

Further reading for those interested:

See [MDN docs prescribing ECMA class initialization order](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#evaluation_order). The thing to note is that in ECMA-compliant classes the constructor _will_ be evaluated before any class field initializers.

See [useDefineForClassFields docs](https://www.typescriptlang.org/tsconfig#useDefineForClassFields). The only useful information here is the default values.

See [original useDefineForClassFields announcement for TS 3.7](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier). This is rather outdated since with target >= ES2022, it does _not_ use Object.defineProperty... It just emits an ECMA class. The functional difference today is that if you opt in to the old behavior, TS will put class field initializers in the emitted constructor _before_ any code that you've written in the constructor, meaning those initializers will run first. And, if you're truly masochistic, you can make it that `this.foo !== foo`.
